openmsupply/open-msupply
# Barcode: lookup item

Fixes #1524

# 👩🏻‍💻 What does this PR do? 
 
Allow a user to scan items when populating an outbound shipment.

- [ ] Add a button to outbounds ( if a scanner is found ) which starts and stops auto scan
- [ ] On a successful scan result, open the 'add item' modal with the item selected
- [ ] On an unsuccessful result - show a warning and open the add item modal without selecting an item 

There's another issue to come which will hook up the android code, save barcodes back and highlight specific batches.

This has required more work than expected; some of the extra work is:
- adding a barcode lookup query, which allows filtering by the code
- refactoring the barcode scanning to allow for callback events, this is so that the user can enable scanning and keep scanning / adding items

# 🧪 How has/should this change been tested? 

You'll need to start electron. For development `yarn electron:start-local` (after running the server and running 'yarn start-local') or more usually `yarn electron:start` though for that, you'll have to build f/e and then start your server, in order to pickup all the f/e changes here.

You can then
- open an outbound; check that there's a scan button
- when clicking scan, it changes the icon and the text to 'stop'
- clicking 'stop' will stop
- navigating away from the outbound will also stop ( save, cancel, reload, and navigation )
- when scanning, a scan result will pop up the add item window
  - with the item selected, if there's a match ( you'll need the barcode in your barcode table )
  - with the item autocomplete open otherwise
- the edit stock line functionality, where you can scan barcodes, should work as normal
- Open the web version, and check that there's no scan button

To manually add a barcode, you can scan an item in the edit-stock-line modal, copy the result and add an entry to your barcode table. You'll need an item id, and will have to copy the barcode value to the `value` column.

You can then check that using the graphQL api too.
here's how it looks:

<img width="675" alt="Screenshot 2023-04-18 at 3 25 53 PM" src="https://user-images.githubusercontent.com/9192912/232970135-55ceb551-0fb0-479c-b5bf-803e32a071aa.png">

https://user-images.githubusercontent.com/9192912/232927445-d2429e9b-e683-4ca0-a22e-17ae343c4f95.mp4


## 💌 Any notes for the reviewer?

There are two USB scanners in the office - you'll need one to test fully
To make the continuous scan work, I enabled on the scanner:

![image](https://user-images.githubusercontent.com/9192912/232931044-f557b05a-2fed-4abc-9ca8-9365a781f985.png)

and they also recommend enabling picklist mode ("In this mode, you can pick out and decode a bar code from a group of bar codes that are printed close together by placing the aiming pattern on the bar code you want to decode.")

![image](https://user-images.githubusercontent.com/9192912/232931102-5d618652-8d90-4a0e-a38a-5ebad0cfe7b1.png)


## 📃 Documentation
- [ ] Barcode scanning in outbounds
